drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2126 @@
+           * After this operation, only the latest ate is valid.
+           * Expire the old one only if it is not deleted before(Or it is
+           * already expired)
+           */
+
+          if (prev_found && wlk_ate.expired == fs->erasestate)
+            {
+              finfo("prev entry exists, expire it\n");
+
+              /* If gc touched second latest ate, search for it again */
+
+              if (gc_count >= fs->geo.neraseblocks - 1 -
+                  (block_to_write_befor_gc + fs->geo.neraseblocks -
+                  (hist_addr >> ADDR_BLOCK_SHIFT))
+                  % fs->geo.neraseblocks)
+                {
+                  rc = nvs_read_hist(fs, key, key_size, NULL, 0, 1,
+                                     &second_latest_ate_addr);
+                  finfo("search for prev entry, %" PRIx32 ", "
+                        "rc %d\n",
+                        second_latest_ate_addr, rc);
+                  if (rc > 0)
+                    {
+                      /* Set second latest ate id to all 0
+                       * so that only the latest ate is valid.
+                       */
+
+                      rc = nvs_expire_ate(fs, second_latest_ate_addr);
+                      if (rc < 0)
+                        {
+                          ferr("expire ate failed, addr %" PRIx32 "\n",
+                            second_latest_ate_addr);
+                          return rc;
+                        }
+                    }
+                }
+              else
+                {
+                  /* Set second latest ate id as old entry
+                   * so that only the latest ate is valid.
+                   */
+
+                  rc = nvs_expire_ate(fs, hist_addr);
+                  finfo("expir prev entry, %" PRIx32 ", rc %d\n",
+                        hist_addr, rc);
+
+                  if (rc < 0)
+                    {
+                      ferr("expire ate failed, addr %" PRIx32 "\n",
+                           hist_addr);
+                      return rc;
+                    }
+                }
+            }
+          break;
+        }
+
+      rc = nvs_block_close(fs);
+      if (rc)
+        {
+          return rc;
+        }
+
+      rc = nvs_gc(fs);
+      if (rc)
+        {
+          return rc;
+        }
+
+      gc_count++;
+      finfo("Gc count=%d\n", gc_count);
+    }
+
+  finfo("nvs_write success\n");
+  return 0;
+}
+
+/****************************************************************************
+ * Name: nvs_delete
+ *
+ * Description:
+ *   Delete an entry from the file system.
+ *
+ * Input Parameters:
+ *   fs    - Pointer to file system.
+ *   pdata - Pointer to data buffer.
+ *
+ * Returned Value:
+ *   0 on success, -ERRNO errno code if error.
+ *
+ ****************************************************************************/
+
+static int nvs_delete(FAR struct nvs_fs *fs, FAR struct config_data_s *pdata)
+{
+  return nvs_write(fs, pdata);
```

Comment:
do use need set pdata->len = 0 before call nvs_write

Reply:
If user wants to use mtd_config_fs driver directly, user needs to set pdata->len=0.
Maybe it is better we set pdata->len =0 in nvs_delete, so that user doesn't need to do that?

Reply:
But, mtdconfig_s doesn't use pdata->len in mtdconfig_deleteconfig, we need keep the consistent behavior, otherwise apps/system/cfgdata/ will stop work.

Reply:
But cfgdatacmd_unset has the following code:
  cfg.configdata = NULL;
  cfg.len = 0;
So, len=0 is used in apps/system/cfgdata?

Reply:
if so, ok.